### PR TITLE
[MNG-8059] Paths everywhere

### PR DESCRIPTION
**Pull request by cstamas**
Make execution flow never invoke path.toFile, at least what model-builder and maven-resolver-provider matters.

---

https://issues.apache.org/jira/browse/MNG-8059

### CODE REVIEW

File: maven-core/src/main/java/org/apache/maven/bridge/MavenRepositorySystem.java [hunk 1]
```diff
@@ -653,7 +653,7 @@ public ArtifactRepository createLocalRepository(MavenExecutionRequest request, F
 
     public ArtifactRepository createLocalRepository(File localRepository) throws InvalidRepositoryException {
         return createRepository(
-                "file://" + localRepository.toURI().getRawPath(),
+                localRepository.toURI().toASCIIString(),
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/bridge/MavenRepositorySystem.java:656]: I hate it with a passion that Sun failed back then to make `#toString()` correctly.

Comment (cstamas) [reply maven-core/src/main/java/org/apache/maven/bridge/MavenRepositorySystem.java:656]: Hej! Could you help me with this? To me, it seems even Maven master works with "wrong" URIs, but somehow `File` works with them (the `new File(uri)` does the job), but to me somehow it looks like we end up with File instance that explodes when `File.toPath()` is invoked on that File instance.

See this PR:
https://github.com/apache/maven/pull/1416

It passes on Win.

Then this PR, where the _only_ diff is that `new File(uri).toPath()` is invoked, and we end up with UT errors on Win like these:
https://gist.github.com/cstamas/8cf6d3e6d55f62c70f7b857effcb7bc2

Comment (cstamas) [reply maven-core/src/main/java/org/apache/maven/bridge/MavenRepositorySystem.java:656]: If you are on Win, you should reproduce this just by checking out this PR, as Resolver on this PR is 'locked snapshot' so you would get the very same thing that CI builds.

Comment (michael-o) [reply maven-core/src/main/java/org/apache/maven/bridge/MavenRepositorySystem.java:656]: > Hej! Could you help me with this? To me, it seems even Maven master works with "wrong" URIs, but somehow `File` works with them (the `new File(uri)` does the job), but to me somehow it looks like we end up with File instance that explodes when `File.toPath()` is invoked on that File instance.
> 
> See this PR: #1416
> 
> It passes on Win.
> 
> Then this PR, where the _only_ diff is that `new File(uri).toPath()` is invoked, and we end up with UT errors on Win like these: https://gist.github.com/cstamas/8cf6d3e6d55f62c70f7b857effcb7bc2

There are subtile conversion issues with `File#toURI()` and `Path#toUri()`. They behave differently. See https://issues.apache.org/jira/browse/SUREFIRE-2211

Try: `localRepository.toPath().toUri().toASCIIString()`. Then check whether it is absolute or that causes a problem.

Comment (michael-o) [reply maven-core/src/main/java/org/apache/maven/bridge/MavenRepositorySystem.java]: All tests pass for me on Windows.

**Comment by michael-o** [comment]
@cstamas Seems my idea does work?

**Comment by cstamas** [comment]
@michael-o unsure, did you change anything that is in this PR? Maybe drive letter matters? As CI keeps talking about `D:`... I mean, why does the CI fail here if all ok?

**Comment by michael-o** [comment]
> @michael-o unsure, did you change anything that is in this PR? Maybe drive letter matters? As CI keeps talking about `D:`... I mean, why does the CI fail here if all ok?

Didn't touch a bit. For me, your chagne runs locally with Java 11.

**Comment by cstamas** [comment]
Hm, I think CI uses Java 11 as well? :shrug:

**Comment by cstamas** [comment]
This happened in resolver: it could allow now directory (classes in a directory) not only JAR files on "system" path:
```
MavenITmng3288SystemScopeDirTest.testitMNG3288:52->AbstractMavenIntegrationTestCase.fail:387 Usage of directory instead of file for system-scoped dependency did not fail dependency
```

File: maven-builder-support/src/main/java/org/apache/maven/building/UrlSource.java [hunk 2]
```diff
@@ -87,6 +87,6 @@ public boolean equals(Object obj) {
         }
 
         UrlSource other = (UrlSource) obj;
-        return this.url.equals(other.url);
+        return Objects.equals(url.toExternalForm(), other.url.toExternalForm());
```
Comment (michael-o) [inline maven-builder-support/src/main/java/org/apache/maven/building/UrlSource.java:90]: Why actually?

Comment (cstamas) [reply maven-builder-support/src/main/java/org/apache/maven/building/UrlSource.java:90]: url equality check performs hostname resolution

Comment (michael-o) [reply maven-builder-support/src/main/java/org/apache/maven/building/UrlSource.java:90]: Really? WTF?! Why not `#toAsciiString()`?

Comment (cstamas) [reply maven-builder-support/src/main/java/org/apache/maven/building/UrlSource.java:90]: Whatever, am just creating string out of both in same way and comparing the two.

File: maven-core/src/main/java/org/apache/maven/bridge/MavenRepositorySystem.java [hunk 3]
```diff
@@ -654,7 +654,7 @@ public ArtifactRepository createLocalRepository(MavenExecutionRequest request, F
 
     public ArtifactRepository createLocalRepository(File localRepository) throws InvalidRepositoryException {
         return createRepository(
-                "file://" + localRepository.toURI().getRawPath(),
+                localRepository.toPath().toUri().toASCIIString(),
```
Comment (cstamas) [inline maven-core/src/main/java/org/apache/maven/bridge/MavenRepositorySystem.java:657]: This change seems unrelated, will try to undo this, as IMHO is actually not needed.

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/MavenSessionBuilderSupplier.java [hunk 4]
```diff
@@ -77,7 +78,7 @@ protected DependencyTraverser getDependencyTraverser() {
     }
 
     protected DependencyManager getDependencyManager() {
-        return new ClassicDependencyManager(true); // same default as in Maven4
+        return new ClassicDependencyManager(true, SystemScopeHandler.LEGACY); // same default as in Maven4
```
Comment (cstamas) [inline maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/MavenSessionBuilderSupplier.java:81]: This is TODO, but to me seems wrong: "same as in maven4, then legacy"?

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/type/DefaultType.java [hunk 5]
```diff
@@ -66,6 +66,13 @@ public DefaultType(
         this.properties = Collections.unmodifiableMap(properties);
     }
 
+    /**
+     * Copy constructor.
+     */
+    public DefaultType(String id) {
```
Comment (cstamas) [inline maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/type/DefaultType.java:72]: TODO: this may be not needed, there is only one use of it AFAIR

File: pom.xml [hunk 6]
```diff
@@ -624,6 +624,20 @@ under the License.
...
+        <enabled>true</enabled>
+        <updatePolicy>never</updatePolicy>
+      </releases>
+      <snapshots>
+        <enabled>false</enabled>
+      </snapshots>
+      <id>maven-2066</id>
+      <url>https://repository.apache.org/content/repositories/maven-2066/</url>
+    </repository>
+  </repositories>
```
Comment (cstamas) [inline pom.xml:640]: Remove once 2.0.0-alpha-8 released.

**Comment by jira-importer** [comment]
Resolve #10809
